### PR TITLE
Send sync implementation on iterators

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #22709.
cc @Veedrac

r? @bluss

### CODE REVIEW

**Comment by bluss** [comment]
We'd only need Send + Sync on `table::RawBuckets`, and the rest should follow by itself. Is that feasible @Gankro?

I had another question for you, do you know why `RawBucket` [is using `PhantomData<(K, V)>`](https://github.com/GuillaumeGomez/rust/blob/send_sync/src/libstd/collections/hash/table.rs#L80-L85). I don't think it owns a `K`, but I don't have it all clear like you.

**Comment by Gankro** [comment]
@bluss it ptr::read's out K,V pairs. Not sure if that necessitates it... this code has seen a lot of changes in the Phantom story.

RawBuckets should... basically be fine to impl for? Doesn't really matter since it's internal.

**Comment by GuillaumeGomez** [comment]
So... What should I add/remove/update ?

**Comment by bluss** [comment]
@Gankro It matters if all the things using RawBuckets would automatically inherit Send + Sync

**Comment by Gankro** [comment]
Nothing in hashmap uses shared state/internal mutability so it's all trivially thread-safe, as far as I know.

**Comment by bluss** [comment]
@GuillaumeGomez If you impl Send + Sync for RawBuckets instead, all the rest should be inherited automatically.

**Comment by bluss** [comment]
Can you add the iterators to sync-send-iterators-in-libcollections.rs? (Or are they somewhere else?)

**Comment by GuillaumeGomez** [comment]
@bluss: sure !

**Comment by GuillaumeGomez** [comment]
I added the test.

**Comment by bluss** [comment]
Thanks! I'd prefer you squash the commits together to clean it up.

r=me with squash

File: src/libstd/collections/hash/table.rs [hunk 1]
```diff
@@ -741,6 +741,9 @@ struct RawBuckets<'a, K, V> {
     marker: marker::PhantomData<&'a ()>,
 }
 
+unsafe impl<'a, K: Send, V: Send> Send for RawBuckets<'a, K, V> {}
```
Comment (huonw) [inline src/libstd/collections/hash/table.rs]: Hm, it's not obvious to me that this is the right thing. Maybe this should be `K: Sync, V: Sync`? Basically, I'm asking what reference/value semantics does `RawBuckets` have with `K` and `V`?

**Comment by bluss** [comment]
Giving this issue to huon, who is much more experienced in the details of correct `Sync` and `Send`.

**Comment by bluss** [comment]
So, my plan make this easier by impl on `RawBuckets` was misguided, and this is because the different iterators do actually need different bounds. The `Send`/`Sync` impl for RawBuckets is not correct -- its behavior depends on use.

For reference, see [this comment](https://github.com/rust-lang/rust/issues/22709#issuecomment-75714311) for the different cases. The guidelines should be directly applicable since the `HashMap` does not do any funny business with their contents during iteration.

We only need to impl `Send / Sync` for the basic building blocks. In this case this is `table::{Iter, IterMut, IntoIter, Drain}` inside `src/libstd/collections/hash/table.rs`. The `HashMap` and `HashSet` iterators will inherit `Send / Sync` from these.

- `impl<'a, K: Sync, V: Sync> Sync for Iter<'a, K, V>`
- `impl<'a, K: Sync, V: Sync> Send for Iter<'a, K, V>`
- `impl<'a, K: Sync, V: Sync> Sync for IterMut<'a, K, V>`
- `impl<'a, K: Send, V: Send> Send for IterMut<'a, K, V>`
- `impl<K: Sync, V: Sync> Sync for IntoIter<K, V>`
- `impl<K: Send, V: Send> Send for IntoIter<K, V>`
- `impl<'a, K: Sync, V: Sync> Sync for Drain<'a, K, V>`
- `impl<'a, K: Send, V: Send> Send for Drain<'a, K, V>`

File: src/libstd/collections/hash/set.rs [hunk 2]
```diff
@@ -764,18 +764,27 @@ pub struct Iter<'a, K: 'a> {
     iter: Keys<'a, K, ()>
 }
 
+unsafe impl<'a, K: Sync> Sync for Iter<'a, K> {}
+unsafe impl<'a, K: Send> Send for Iter<'a, K> {}
```
Comment (huonw) [inline src/libstd/collections/hash/set.rs]: This needs to be `K: Sync`, since `Iter` yields `&K`.

File: src/libstd/collections/hash/table.rs [hunk 3]
```diff
@@ -818,6 +818,9 @@ pub struct Iter<'a, K: 'a, V: 'a> {
     elems_left: usize,
 }
 
+unsafe impl<'a, K: Sync, V: Sync> Sync for Iter<'a, K, V> {}
+unsafe impl<'a, K: Send, V: Send> Send for Iter<'a, K, V> {}
```
Comment (huonw) [inline src/libstd/collections/hash/table.rs]: These both need to be `: Sync`.

File: src/libstd/collections/hash/table.rs [hunk 4]
```diff
@@ -835,18 +838,27 @@ pub struct IterMut<'a, K: 'a, V: 'a> {
     elems_left: usize,
 }
 
+unsafe impl<'a, K: Sync, V: Sync> Sync for IterMut<'a, K, V> {}
+unsafe impl<'a, K: Send, V: Send> Send for IterMut<'a, K, V> {}
```
Comment (huonw) [inline src/libstd/collections/hash/table.rs]: This should be `K: Sync`, `V: Send`.

Comment (huonw) [inline src/libstd/collections/hash/table.rs]: Actually, after thinking some more, this isn't so obvious. The implementation is `Iterator<Item = (&K, &mut V)>`, so "naively" the `K` requires `Sync`, however, in practice the `&K` is guaranteed to be unaliased (it's "really" `&uniq K`) so maybe `K: Send` works here.

I'm a bit confused about the correctness here. I'm now thinking that either `K: Sync` or `K: Send` are correct, but that seems mildly strange to me. I'll ask.

File: src/libstd/collections/hash/set.rs [hunk 5]
```diff
@@ -764,18 +764,27 @@ pub struct Iter<'a, K: 'a> {
     iter: Keys<'a, K, ()>
 }
 
+unsafe impl<'a, K: Sync> Sync for Iter<'a, K> {}
+unsafe impl<'a, K: Sync> Send for Iter<'a, K> {}
```
Comment (alexcrichton) [inline src/libstd/collections/hash/set.rs]: Can these be automatically inferred by applying the traits to the underlying `Keys` iterator?

Comment (alexcrichton) [inline src/libstd/collections/hash/set.rs]: Er actually, all of the modifications in this file should not be necessary as they're built on top of the map iterators

File: src/libstd/collections/hash/table.rs [hunk 6]
```diff
@@ -835,18 +838,27 @@ pub struct IterMut<'a, K: 'a, V: 'a> {
     elems_left: usize,
 }
 
+unsafe impl<'a, K: Sync, V: Sync> Sync for IterMut<'a, K, V> {}
+unsafe impl<'a, K: Sync, V: Send> Send for IterMut<'a, K, V> {}
```
Comment (alexcrichton) [inline src/libstd/collections/hash/table.rs]: Hm, @huonw do you think this needs `Sync` on `K` for the iter mut case? If you have an `IterMut` you're guaranteed you have the only reference to the map, so each reference handed out to the keys should be unique, right?

Comment (huonw) [inline src/libstd/collections/hash/table.rs]: I'm unsure. See https://github.com/rust-lang/rust/pull/27615#discussion_r36679199 . We're also discussing it on IRC at the moment:

- https://botbot.me/mozilla/rust-libs/msg/46882011/
- https://botbot.me/mozilla/rust-internals/msg/46880251/

I think the conclusion is that both `K: Send` and `K: Sync` are correct, and that `K: Send` is the more useful bound (i.e. there are more things that are `Send` but not `Sync` than visa versa).

Comment (alexcrichton) [inline src/libstd/collections/hash/table.rs]: That sounds like the conclusion I came to as well!

Comment (bluss) [inline src/libstd/collections/hash/table.rs]: @GuillaumeGomez As you can see this is tricky stuff! Can you change `IterMut` to use `K: Send` for its `Send` impl? I'd add a comment too, maybe like this:

```
// Both K: Sync and K: Send are correct for IterMut's Send impl, but Send is the more useful bound
```

Comment (GuillaumeGomez) [inline src/libstd/collections/hash/table.rs]: Yes of course. For the squash, you'll have to wait this evening. ^^

**Pull request merged**